### ISSUE TITLE
Public key information is not mandatory for ECDSA private keys

### ISSUE THREAD

**Issue by YuryStrozhevsky** [PeculiarVentures/node-webcrypto-ossl#123]
When I am trying to import private key via this code I have error:
```
TypeError: Cannot read property 'length' of undefined
    at Function.Base64Url.decode (node_modules\webcrypto-core\dist\webcrypto-core.js:275:26)
    at b64_decode (node_modules\node-webcrypto-ossl\buildjs\crypto\ec.js:38:33)
    at node_modules\node-webcrypto-ossl\buildjs\crypto\ec.js:123:33
    at new Promise (<anonymous>)
    at Function.EcCrypto.importKey (node_modules\node-webcrypto-ossl\buildjs\crypto\ec.js:75:16)
    at node_modules\node-webcrypto-ossl\buildjs\subtle.js:363:29
    at <anonymous>
```

```javascript
import { getCrypto, setEngine } from "../src/common";
import CryptoEngine from "../src/CryptoEngine";

const WebCrypto = require("node-webcrypto-ossl");
const webcrypto = new WebCrypto();

setEngine("newEngine", webcrypto, new CryptoEngine({ name: "", crypto: webcrypto, subtle: webcrypto.subtle }));

async function main()
{
 const crypto = getCrypto();
 const importedPrivateKey = await crypto.importKey("jwk", { crv: "P-256", kty: "EC", d: "3BVP-DHyzDs1o0AFLS9PNBRUVLAFw8cMmY7w1VFgkGY"}, { name: "ECDSA", namedCurve: "P-256" }, true, ["sign"]);
}

main();
```
At the same time native WebCrypto could successfully import this key.

The problem in your code is that you presume `x` and `y` values are always available inside `jwk`. All these values are part of ECPublicKey type. But in fact for ECPrivate key we have this ASN.1 schema:
```
 ECPrivateKey ::= SEQUENCE {
 version        INTEGER { ecPrivkeyVer1(1) } (ecPrivkeyVer1),
 privateKey     OCTET STRING,
 parameters [0] ECParameters {{ NamedCurve }} OPTIONAL,
 publicKey  [1] BIT STRING OPTIONAL
 }
```
So, as you can see `publicKey` is optional and could be absent inside ECPrivateKey declaration.

**Comment by microshine** [comment]
Which native crypto do you use for test?

I ran code in Chrome, Firefox, and Safari
```js
crypto.subtle.importKey("jwk", { crv: "P-256", kty: "EC", d: "3BVP-DHyzDs1o0AFLS9PNBRUVLAFw8cMmY7w1VFgkGY" }, { name: "ECDSA", namedCurve: "P-256" }, true, ["sign"])
    .then((k) => console.log(k))
    .catch((err) => console.error(err));
```
### Chrome
```
DOMException: The required JWK member "x" was missing
```
### Safari
```
DataError (DOM Exception 30): Data provided to an operation does not meet requirements
```

### Firefox
```
DOMException: Data provided to an operation does not meet requirements
```

**Comment by YuryStrozhevsky** [comment]
No problem in Chrome 65.0.3325.181:
![untitled_2](https://user-images.githubusercontent.com/6128446/38249967-1f5e4926-3756-11e8-8206-f070920dfff9.jpg)

I will try to run at another browsers soon.

**Comment by microshine** [comment]
I use the same version of Chrome
My OS is MacOS High Sierra 10.13.3

![image](https://user-images.githubusercontent.com/7333669/38250030-57504eb0-3756-11e8-92aa-78a1546846bc.png)

Here is chromium source code. They use `x` and `y` for private key creation
https://cs.chromium.org/chromium/src/components/webcrypto/algorithms/ec.cc?sq=package:chromium&l=500-512

**Comment by YuryStrozhevsky** [comment]
Very interesting - when I am using Web application with native `import` directives all works just fine. But when I am trying to run ES5 application I got DOM Exception for the same code.

Anyway, even if you have DOM Exception it is wrong behavior and seems in latest Chrome code it was already fixed. I do suggest you to fix your library as well - as I said data from public key is only optional for private EC keys.

**Comment by microshine** [comment]
I see. I'll look at this and try to fix

This code throws error in Chrome too
```js
class Test {
  static async run() {
    const key = await crypto.subtle.importKey("jwk", { crv: "P-256", kty: "EC", d: "3BVP-DHyzDs1o0AFLS9PNBRUVLAFw8cMmY7w1VFgkGY" }, { name: "ECDSA", namedCurve: "P-256" }, true, ["sign"]);
  }
}
Test.run();
```

**Comment by microshine** [comment]
@YuryStrozhevsky can you export created key to `jwk` and `pkcs8` and share result?

**Comment by YuryStrozhevsky** [comment]
Hm, seems I had success for page used `webcrypto-liner`. When I removed it I got the same DOM Exception. 

So, here there is a few ways how to process:
1. Aline code between `webcrypto-liner` and `node-webcrypto-ossl`;
2. Try to file a bug to Chrome. I will investigate JWK a little bit deeper and cover this item;

But for now I need to confirm `node-webcrypto-ossl` aligned with existing Chrome native WebCrypto. Closing the issue then.
